scipy/linalg/src/_batched_linalg.h

```diff
@@ -0,0 +1,305 @@
+ * Copy each slice into a buffer, make the buffer F-ordered
+ * for LAPACK.
+ */
+struct iter_data_t
+{
+    // grab useful data from the array to iterate
+    npy_intp ndim;
+    npy_intp *shape;
+    npy_intp *strides;
+    void *data_ptr;
+
+    npy_intp outer_size;  // math.prod(a.shape[:-2])
+    npy_intp m, n;        // core dimensions
+
+    iter_data_t(PyArrayObject *arr) {
+        ndim = PyArray_NDIM(arr);
+        shape = PyArray_SHAPE(arr);
+        strides = PyArray_STRIDES(arr);
+        data_ptr = PyArray_DATA(arr);
+
+        m = shape[ndim - 2];
+        n = shape[ndim - 1];
+
+        /* outer_size: the number of slices in the batch, math.prod(a.shape[:-2] */
+        outer_size = 1;
+        if(ndim > 2) {
+            for(int i=0; i < ndim - 2; i++) {
+                outer_size *= shape[i];
+            }
+        }
+    }
+
+    /*
+     * Copy the slice number `idx` into the `buffer`.
+     */
+    template<typename T>
+    void copy_slice(const npy_intp idx, T *buffer) {
+
+        assert((0 <= idx) && (idx < outer_size));
+
+        /* parroted from sqrtm*/
+        npy_intp offset = 0;
+        npy_intp temp_idx = idx;
+        for (int i = ndim - 3; i >= 0; i--) {
+            offset += (temp_idx % shape[i]) * strides[i];
+            temp_idx /= shape[i];
+        }
+
+        T *slice_ptr = (T *)(data_ptr) + offset/sizeof(T);
+
+        // copy the current n-x-n slice into the temp storage *in the Fortran order*
+        for(npy_intp i=0; i < m; i++) {
+            for(npy_intp j=0; j < n; j++) {
+                buffer[i + j*m] = *(slice_ptr + (i*strides[ndim - 2]/sizeof(T)) + (j*strides[ndim - 1]/sizeof(T)));
+            }
+        }
+    }
+};
+
+
+/***********************************
+ *   Inversion via GETRF / GETRI
+ ***********************************/
+
+/*
+ * Invert a 2D slice:
+ *   - Input slice is in `getrf_data.a`.
+ *   - The result is in `getri_data.a`.
+ *
+ * Return is 0 on success; on failure, return the failing operation's
+ * `info` variable.
+ *
+ */
+template<typename T>
+int
+invert_slice(getrf_data_t<T>& getrf_data, getri_data_t<T>& getri_data) {
+
+    // factorize
+    call_getrf(getrf_data);
+
+    if (getrf_data.info != 0) {
+        return getrf_data.info;
+    }
+
+    // prepare the data for the GETRI call; other getri_data members do not change
+    // between iterations.
+    getri_data.a = getrf_data.a;
+    getri_data.ipiv = getrf_data.ipiv;
+
+    // compute the inverse
+    call_getri(getri_data);
+
+    return getri_data.info;
+}
+
+
+/*
+ * The main batched-inv routine, to be called from the python API layer.
+ *
+ * This is a computational routine, and it assumes that the caller has done all input
+ * validation before calling it.
+ * In particular:
+ *
+ * For the input array `arr`:
+ *
+ * - The caller must check that the dtype is LAPACK-compatible before calling this routine.
+ * - The array is at least 3D (for a single 2D matrix, the caller must
```

Comment:
This is not necessary, you can detect this from the ndim property of the array. Then you loop only once in the outer slice loop if ndim = 2. That's the setting to `1` in sqrtm outer dimension multiplication.